### PR TITLE
Resolve type stability issue in payoff_vector

### PR DESCRIPTION
**Pull request by oyamad**
Implement the approach in https://github.com/QuantEcon/Games.jl/issues/2#issuecomment-244659348
- Replace `_reduce_last_player` with `_reduce_ith_opponent` which now keeps the dimension
- Introduce `PureActionProfile` and `MixedActionProfile`
- Change `ActionProfile` to `Union{PureActionProfile,MixedActionProfile}`
- Declare the type of `payoffs` in `payoff_vector`

### CODE REVIEW

File: src/normal_form_game.jl [hunk 1]
```diff
@@ -188,7 +214,7 @@ valse otherwise.
 """
 function is_best_response(player::Player,
                           own_action::PureAction,
-                          opponents_actions::Union{Action,ActionProfile,Void};
+                          opponents_actions::Union{Action,PureActionProfile,MixedActionProfile,Void};
```
Comment (spencerlyon2) [inline src/normal_form_game.jl]: It seems that we are using this particular `Union` quite often. Does it make sense to name it using `typealias`?

Also, it is true that `Union{A,Union{B,C}} == Union{A,B,C}` so even if we don't name this particular Union, we should be able to make use of `ActionProfile` to replace `PureActionProfile,MixedActionProfile`

Comment (oyamad) [inline src/normal_form_game.jl]: Good point. I replaced `PureActionProfile,MixedActionProfile` with `ActionProfile`.

**Comment by oyamad** [comment]
Can I merge this?

**Comment by cc7768** [comment]
Yes. This looks good to me.

Thanks for cleaning up the type stability.

**Pull request merged**